sdk/src/message.rs

```diff
@@ -84,36 +76,49 @@ pub struct Message {
     /// signatures must match the first `num_required_signatures` of `account_keys`.
     pub num_required_signatures: u8,
 
+    /// The number of credit-only signed accounts. The last `num_credit_only_signed_accounts` of
+    /// the signed keys will be cached during runtime to allow parallel processing across
+    /// transactions.
+    pub num_credit_only_signed_accounts: u8,
+
+    /// The number of credit-only unsigned accounts. The last `num_credit_only_unsigned_accounts`
+    /// of the unsigned keys will be cached during runtime to allow parallel processing across
+    /// transactions.
+    pub num_credit_only_unsigned_accounts: u8,
+
     /// All the account keys used by this transaction
     #[serde(with = "short_vec")]
     pub account_keys: Vec<Pubkey>,
 
     /// The id of a recent ledger entry.
     pub recent_blockhash: Hash,
 
-    /// All the program id keys used to execute this transaction's instructions
-    #[serde(with = "short_vec")]
-    program_ids: Vec<Pubkey>,
-
     /// Programs that will be executed in sequence and committed in one atomic transaction if all
     /// succeed.
     #[serde(with = "short_vec")]
     pub instructions: Vec<CompiledInstruction>,
 }
 
+// The serialized size of the message intro data before account_keys: size of
+// num_required_signatures + size of num_credit_only_signed_accounts + size of
+// num_credit_only_unsigned_accounts.
+pub const MESSAGE_INTRO_BYTES: usize = 3;
```

Comment:
How about moving all these into a new MessageHeader struct and then using `serialized_size` instead of a constant?

Reply:
Yes, I like that.